File: php/src/Resources/Links.php
<?php

declare(strict_types=1);

namespace Rerout\Resources;

use Rerout\Exceptions\ReroutException;
use Rerout\Models\CreateLinkInput;
use Rerout\Models\Link;
use Rerout\Models\LinkStats;
use Rerout\Models\ListLinksResult;
use Rerout\Models\UpdateLinkInput;
use Rerout\Rerout;

/**
 * Link operations namespace. Reached via `Rerout::links()`.
 */
final class Links
{
    public function __construct(private readonly Rerout $client)
    {
    }

    /** Create a new short link. */
    public function create(CreateLinkInput $input): Link
    {
        /** @var array<string, mixed> $response */
        $response = $this->client->request(
            method: 'POST',
            path: '/v1/links',
            body: $input->toArray(),
        );

        return Link::fromArray($response);
    }

    /** Paginated list of links in the project. */
    public function list(?int $cursor = null, ?int $limit = null): ListLinksResult
    {
        /** @var array<string, mixed> $response */
        $response = $this->client->request(
            method: 'GET',
            path: '/v1/links',
            query: ['cursor' => $cursor, 'limit' => $limit],
        );

        return ListLinksResult::fromArray($response);
    }

    /** Get a single link by code. */
    public function get(string $code): Link
    {
        /** @var array<string, mixed> $response */
        $response = $this->client->request(
            method: 'GET',
            path: '/v1/links/' . rawurlencode($code),
        );

        return Link::fromArray($response);
    }

    /**
     * Patch a link. Only fields set on `$input` are sent. Throws
     * `ReroutException` with code `bad_request` when called with an empty
     * input — the API rejects no-op PATCH bodies.
     */
    public function update(string $code, UpdateLinkInput $input): Link
    {
        if ($input->isEmpty()) {
            throw new ReroutException(
                errorCode: 'bad_request',
                message: 'UpdateLinkInput has no fields to send.',
                status: 0,
                path: '/v1/links/' . $code,
            );
        }

        /** @var array<string, mixed> $response */
        $response = $this->client->request(
            method: 'PATCH',
            path: '/v1/links/' . rawurlencode($code),
            body: $input->toArray(),
        );

        return Link::fromArray($response);
    }

    /**
     * Soft-delete a link. The short URL stops redirecting and is gone from
     * lists.
     *
     * @return bool Whether the server confirmed deletion.
     */
    public function delete(string $code): bool
    {
        $response = $this->client->request(
            method: 'DELETE',
            path: '/v1/links/' . rawurlencode($code),
        );

        if (is_array($response) && array_key_exists('deleted', $response)) {
            $deleted = $response['deleted'];

            return is_bool($deleted) ? $deleted : true;
        }

        return true;
    }

    /** Per-link click stats. Defaults to 30 days. */
    public function stats(string $code, int $days = 30): LinkStats
    {
        /** @var array<string, mixed> $response */
        $response = $this->client->request(
            method: 'GET',
            path: '/v1/links/' . rawurlencode($code) . '/stats',
            query: ['days' => $days],
        );

        return LinkStats::fromArray($response);
    }
}

File: dotnet/src/Rerout/Resources/Links.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Net.Http;
using System.Threading;
using System.Threading.Tasks;
using Rerout.Internal;
using Rerout.Models;

namespace Rerout.Resources;

/// <summary>
/// Link operations: create, list, get, update, delete, and per-link stats.
/// Reached through <see cref="ReroutClient.Links"/>.
/// </summary>
public sealed class Links
{
    private readonly ReroutHttpHandler _http;

    internal Links(ReroutHttpHandler http) => _http = http;

    /// <summary>Create a new short link.</summary>
    /// <param name="input">The link to create. <c>TargetUrl</c> is required.</param>
    /// <param name="cancellationToken">Optional cancellation token.</param>
    /// <returns>The created <see cref="Link"/>.</returns>
    /// <exception cref="ReroutException">On any API or transport failure.</exception>
    public Task<Link> CreateAsync(CreateLinkInput input, CancellationToken cancellationToken = default)
    {
        return _http.SendAsync<Link>(
            new ReroutRequest
            {
                Method = HttpMethod.Post,
                Path = "/v1/links",
                Body = input,
            },
            cancellationToken);
    }

    /// <summary>
    /// Create many links in a single request. Returns a partial-success
    /// envelope — inspect each <see cref="BatchLinkResult.Ok"/> for the per-item
    /// outcome.
    /// </summary>
    /// <param name="links">The links to create, in order.</param>
    /// <param name="cancellationToken">Optional cancellation token.</param>
    /// <exception cref="ReroutException">On any API or transport failure.</exception>
    public Task<BatchCreateLinksResult> CreateBatchAsync(
        IReadOnlyList<BatchLinkInput> links,
        CancellationToken cancellationToken = default)
    {
        return _http.SendAsync<BatchCreateLinksResult>(
            new ReroutRequest
            {
                Method = HttpMethod.Post,
                Path = "/v1/links/batch",
                Body = new Dictionary<string, object?> { ["links"] = links },
            },
            cancellationToken);
    }

    /// <summary>List links in the project, newest first.</summary>
    /// <param name="cursor">Pagination cursor from a previous call's <c>NextCursor</c>.</param>
    /// <param name="limit">Page size. Server-side default and maximum apply.</param>
    /// <param name="cancellationToken">Optional cancellation token.</param>
    public Task<ListLinksResult> ListAsync(
        long? cursor = null,
        int? limit = null,
        CancellationToken cancellationToken = default)
    {
        var query = new List<KeyValuePair<string, string>>();
        if (cursor is { } c)
        {
            query.Add(new KeyValuePair<string, string>("cursor", c.ToString(CultureInfo.InvariantCulture)));
        }

        if (limit is { } l)
        {
            query.Add(new KeyValuePair<string, string>("limit", l.ToString(CultureInfo.InvariantCulture)));
        }

        return _http.SendAsync<ListLinksResult>(
            new ReroutRequest
            {
                Method = HttpMethod.Get,
                Path = "/v1/links",
                Query = query,
            },
            cancellationToken);
    }

    /// <summary>Fetch a single link by its short code.</summary>
    public Task<Link> GetAsync(string code, CancellationToken cancellationToken = default)
    {
        return _http.SendAsync<Link>(
            new ReroutRequest
            {
                Method = HttpMethod.Get,
                Path = $"/v1/links/{EncodeCode(code)}",
            },
            cancellationToken);
    }

    /// <summary>
    /// Patch a link. Only fields set on <paramref name="input"/> are sent — see
    /// <see cref="UpdateLinkInput"/> for the omit / set / clear semantics.
    /// </summary>
    /// <exception cref="ReroutException">
    /// Thrown client-side with code <c>empty_update</c> when
    /// <paramref name="input"/> has no fields to send — the request never
    /// reaches the API.
    /// </exception>
    public Task<Link> UpdateAsync(string code, UpdateLinkInput input, CancellationToken cancellationToken = default)
    {
        if (input.IsEmpty)
        {
            throw new ReroutException(
                "empty_update",
                "UpdateLinkInput has no fields to send. Set at least one field.",
                0,
                path: $"/v1/links/{code}");
        }

        return _http.SendAsync<Link>(
            new ReroutRequest
            {
                Method = HttpMethod.Patch,
                Path = $"/v1/links/{EncodeCode(code)}",
                Body = input.ToPayload(),
            },
            cancellationToken);
    }

    /// <summary>
    /// Soft-delete a link. The short URL stops redirecting and drops out of
    /// list results.
    /// </summary>
    public Task<DeleteResult> DeleteAsync(string code, CancellationToken cancellationToken = default)
    {
        return _http.SendAsync<DeleteResult>(
            new ReroutRequest
            {
                Method = HttpMethod.Delete,
                Path = $"/v1/links/{EncodeCode(code)}",
            },
            cancellationToken);
    }

    /// <summary>Per-link click statistics across the requested window.</summary>
    /// <param name="code">The short code.</param>
    /// <param name="days">Window size in days. Defaults to 30.</param>
    /// <param name="cancellationToken">Optional cancellation token.</param>
    public Task<LinkStats> StatsAsync(string code, int days = 30, CancellationToken cancellationToken = default)
    {
        return _http.SendAsync<LinkStats>(
            new ReroutRequest
            {
                Method = HttpMethod.Get,
                Path = $"/v1/links/{EncodeCode(code)}/stats",
                Query = new[]
                {
                    new KeyValuePair<string, string>("days", days.ToString(CultureInfo.InvariantCulture)),
                },
            },
            cancellationToken);
    }

    internal static string EncodeCode(string code) => Uri.EscapeDataString(code);
}
